### PR TITLE
Allow multiple instantions of workflow which use the same model/db.

### PR DESCRIPTION
**Pull request by koparasy**
Extends the databases, surrogate, hdcache to act as a dictionary of singletons. Keys of the dictionary are the paths to the respective model, faiss index etc.

### CODE REVIEW

File: examples/main.cpp [hunk 1]
```diff
@@ -266,12 +272,14 @@ int main(int argc, char **argv)
     dbType = AMSDBType::RMQ;
   }
 
-  AMSUQPolicy uq_policy =
-      (std::strcmp(uq_policy_opt, "max") == 0) ? AMSUQPolicy::FAISSMax: AMSUQPolicy::FAISSMean;
+  AMSUQPolicy uq_policy = (std::strcmp(uq_policy_opt, "max") == 0)
```
Comment (ggeorgakoudis) [inline examples/main.cpp:275]: Is using strcmp instead of == idiomatic?

File: src/wf/basedb.hpp [hunk 2]
```diff
@@ -1303,13 +1334,14 @@ class RabbitMQDB final : public BaseDB<TypeValue>
...
 
-    _sender = std::make_shared<struct rmq_sender>(); 
+    _sender = std::make_shared<struct rmq_sender>();
     _sender->loop = _loop_sender;
     _evbuffer = new EventBuffer<TypeValue>(_rank,
                                            _loop_sender,
                                            _channel_send,
                                            _queue_sender);
-    if (pthread_create(&_sender->id, NULL, start_worker_sender, _sender.get())) {
+    if (pthread_create(
```
Comment (ggeorgakoudis) [inline src/wf/basedb.hpp:1343]: Why not std::thread?

File: src/wf/utils.hpp [hunk 3]
```diff
@@ -56,6 +55,16 @@ void random_uq_host(bool *uq_flags, int ndata, double acceptable_error)
...
+inline bool is_real_equal(double l, double r)
+{
+  return r == std::nextafter(l, r);
+}
+
+inline bool is_real_equal(float l, float r)
+{
+  return r == std::nextafter(l, r);
+}
+
```
Comment (ggeorgakoudis) [inline src/wf/utils.hpp]: Are those used?

File: src/wf/workflow.hpp [hunk 4]
```diff
@@ -229,11 +232,11 @@ class AMSWorkflow
...
   {
     DBG(Workflow, "Destroying Workflow Handler");
-    if (hdcache) delete hdcache;
+    //    if (hdcache) delete hdcache;
 
-    if (surrogate) delete surrogate;
+    //    if (surrogate) delete surrogate;
 
-    if (DB) delete DB;
+    //    if (DB) delete DB;
```
Comment (ggeorgakoudis) [inline src/wf/workflow.hpp]: Remove

File: src/wf/workflow.hpp [hunk 5]
```diff
@@ -291,17 +294,22 @@ class AMSWorkflow
...
+           outputDim);
     // To move around the inputs, outputs we bundle them as std::vectors
     std::vector<const FPTypeValue *> origInputs(inputs, inputs + inputDim);
     std::vector<FPTypeValue *> origOutputs(outputs, outputs + outputDim);
 
     REPORT_MEM_USAGE(Workflow, "Start")
 
-    if ( surrogate == nullptr ){
-      FPTypeValue **tmpInputs = const_cast<FPTypeValue**>(inputs);
+    if (surrogate == nullptr) {
```
Comment (ggeorgakoudis) [inline src/wf/workflow.hpp]: IS this idiomatic? Why not `if (!surrogate)`

File: src/wf/workflow.hpp [hunk 6]
```diff
@@ -369,38 +376,36 @@ class AMSWorkflow
     }
 
     {
-      void** iPtr = reinterpret_cast<void **>(packedInputs.data());
-      void** oPtr = reinterpret_cast<void **>(packedOutputs.data());
+      void **iPtr = reinterpret_cast<void **>(packedInputs.data());
```
Comment (ggeorgakoudis) [inline src/wf/workflow.hpp:374]: Are those formatting changes only? Is the incoming format the right one? (applies to other code changes too)

File: src/wf/basedb.hpp (hunk 2 above)
Comment (koparasy) [reply src/wf/basedb.hpp:1343]: This is a formatting change. @lpottier why aren't we using std::thread? Is there a limitation coming from RMQ?

File: src/wf/workflow.hpp (hunk 6 above)
Comment (koparasy) [reply src/wf/workflow.hpp:374]: Those are the ones from our .clang-format file. Should be correct.

File: src/ml/hdcache.hpp [hunk 7]
```diff
@@ -150,17 +156,114 @@ class HDCache
...
+        throw std::runtime_error(
+            "We do not support caches of different number of neighbors.");
+
+      // FIXME: Here we need to cast both to float. FAISS index only works for
+      // single precision and we shoehorn FAISS inability to support arbitary real
+      // types by forcing TypeValue to be 'float'. In our case this results in having
+      // cases where input data are of type(TypeInValue) double. Thus here, threshold can
+      // be of different type than 'acceptable_error' and at compile time we cannot decide
+      // which overloaded function to pick.
+      if (!is_real_equal(static_cast<float>(threshold),
```
Comment (koparasy) [inline src/ml/hdcache.hpp:191]: @ggeorgakoudis  here we use these functions.

File: src/wf/basedb.hpp (hunk 2 above)
Comment (lpottier) [reply src/wf/basedb.hpp:1343]: @koparasy @ggeorgakoudis No limitations that I am aware of. Libevent uses Pthread but I could use `std::thread` here. It's a matter of style (pure C++ versus C-ish), also pthread supports signals (impossible to send a signal to a thread with `std::thread`) and I am using that. However, I am thinking to moving to `std::thread` and use conditional variables to mimic signals.

Comment (ggeorgakoudis) [reply src/wf/basedb.hpp:1343]: @lpottier Can you point and explain the use of signals? It'd be good to review that for this recommendation.

Comment (koparasy) [reply src/wf/basedb.hpp:1343]: @ggeorgakoudis can we take this discussion on slack? The modification is a formatting one and orthogonal to this PR. If you don't have any other concerns please accept the PR.

Comment (ggeorgakoudis) [reply src/wf/basedb.hpp:1343]: Agreed, I'll resolve this conversation, 3 more to go

File: src/wf/workflow.hpp (hunk 5 above)
Comment (koparasy) [reply src/wf/workflow.hpp]: Fixed

File: src/wf/workflow.hpp (hunk 4 above)
Comment (koparasy) [reply src/wf/workflow.hpp]: Fixed

File: examples/main.cpp (hunk 1 above)
Comment (koparasy) [reply examples/main.cpp:275]: Fixed

**Review by ggeorgakoudis — APPROVED** [review]

**Pull request merged**